llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp

```diff
@@ -1919,27 +1919,14 @@ static Value *optimizeTrigReflections(CallInst *Call, LibFunc Func,
   // TODO: Can this be shared to also handle LLVM intrinsics?
   Value *X;
   switch (Func) {
-  case LibFunc_sin:
-  case LibFunc_sinf:
-  case LibFunc_sinl:
-  case LibFunc_tan:
-  case LibFunc_tanf:
-  case LibFunc_tanl:
-    // sin(-X) --> -sin(X)
-    // tan(-X) --> -tan(X)
-    if (match(Call->getArgOperand(0), m_OneUse(m_FNeg(m_Value(X)))))
-      return B.CreateFNeg(
-          copyFlags(*Call, B.CreateCall(Call->getCalledFunction(), X)));
-    break;
   case LibFunc_cos:
   case LibFunc_cosf:
   case LibFunc_cosl: {
-    // cos(-x) --> cos(x)
     // cos(fabs(x)) --> cos(x)
     // cos(copysign(x, y)) --> cos(x)
     Value *Sign;
     Value *Src = Call->getArgOperand(0);
-    if (match(Src, m_FNeg(m_Value(X))) || match(Src, m_FAbs(m_Value(X))) ||
+    if (match(Src, m_FAbs(m_Value(X))) ||
```

Comment:
You can wrap the `fabs` and `copysign` stuff into the symmetric call. e.g., note that `f(fabs(x))` is `f(x)` if `f` is even and `fabs(f(x))` if `f` is odd. That would eliminate this method entirely.

Reply:
Done ✅

I've decided to mimic the existing functionality, so do it only for the even case (where it has a clear win).